clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -217,20 +217,28 @@ public RecordAppendResult append(TopicPartition tp,
         int count = 0;
         for (Map.Entry<TopicPartition, Deque<RecordBatch>> entry : this.batches.entrySet()) {
             Deque<RecordBatch> dq = entry.getValue();
-            synchronized (dq) {
-                // iterate over the batches and expire them if they have stayed in accumulator for more than requestTimeOut
-                Iterator<RecordBatch> batchIterator = dq.iterator();
-                while (batchIterator.hasNext()) {
-                    RecordBatch batch = batchIterator.next();
-                    // check if the batch is expired
-                    if (batch.maybeExpire(requestTimeout, now, this.lingerMs)) {
-                        expiredBatches.add(batch);
-                        count++;
-                        batchIterator.remove();
-                        deallocate(batch);
-                    } else {
-                        if (!batch.inRetry()) {
-                            break;
+            // Only expire a batch if its metadata is not available.
+            if (cluster.leaderFor(entry.getKey()) == null) {
+                synchronized (dq) {
+                    // iterate over the batches and expire them if they have stayed in accumulator for more than requestTimeOut
+                    RecordBatch lastBatch = dq.peekLast();
+                    Iterator<RecordBatch> batchIterator = dq.iterator();
+                    while (batchIterator.hasNext()) {
+                        RecordBatch batch = batchIterator.next();
+                        // We have to close the previous batches in the deque before expiration check, otherwise those
+                        // batches may never be considered as full.
+                        if (batch != lastBatch)
+                            batch.records.close();
```

Comment:
@junrao If we don't close the batch here. What could happen is that a batch still has a few bytes free space left, but it cannot append more messages. But when we call records.isFull() it will return false even if the batch is actually ready. 
In the RecordAccumulator.drain() code, a full batch is determined by either either deque size is > 1 or records.isFull() == true. I am applying the same logic here.